docs/_ext/autoclass_with_examples.py

```diff
@@ -0,0 +1,108 @@
+from sphinx.ext.autodoc.directive import AutodocDirective
+
+from starknet_py.constants import ROOT_PATH
+
+
+class AutoclassWithExamples(AutodocDirective):
+    """
+    Custom extension of the AutodocDirective class, which is used to
+    add code examples to the method docstrings in the documentation.
+
+    This class runs before the AutodocDirective class and pulls code snippets
+    from the starknet_py/tests/e2e/docs/code_examples directory to include in the documentation.
+    This allows developers to easily see and understand how the methods
+    being documented are intended to be used in a practical context.
+    """
+
+    def run(self) -> List[Node]:
+        original_class = getattr(
+            sys.modules[self.env.ref_context.get("py:module")], self.arguments[0]
+        )
+        add_code_examples(original_class)
+
+        self.name = self.name.replace("-with-examples", "")  # remove `-with-examples`
+
+        return AutodocDirective.run(self)
+
+
+def add_code_examples(original_class: Any):
+    """
+    Adds code examples for the given class.
+    """
+    base_class = original_class.__base__
+
+    file_name, file_content = _extract_file_properties(original_class.__name__)
+
+    for method_name, method in original_class.__dict__.items():
+        if not callable(method):
+            continue
+
+        stripped_method_name = method_name.strip("_")
+        if _code_example_exists(stripped_method_name, file_content):
+            hint = _create_hint(file_name, stripped_method_name)
+            _append_hint(method_name, method, base_class, hint)
+
+
+def _append_hint(method_name: str, method, base_class: Any, hint: str):
+    """
+    If method does not have the __doc__, takes it from the base method.
+    """
+    if method.__doc__ is None and method_name in base_class.__dict__:
+        method.__doc__ = getattr(base_class, method_name).__doc__ + hint
+    else:
+        method.__doc__ = (method.__doc__ or "") + hint
+
+
+def _extract_file_properties(class_name: str) -> Tuple[str, str]:
+    """
+    Extracts file content for given class name.
+
+    :param class_name: A string representing the name of the class where examples will be added to.
+    :returns: A tuple containing the name of the file as a string, and its content as a string.
+    """
+    file_name = "test_" + _camel_to_snake(class_name) + ".py"
+    file_path = ROOT_PATH / "tests/e2e/docs/code_examples" / file_name
+
+    with open(file_path, "r", encoding="utf-8") as file:
+        return file_name, file.read()
+
+
+def _camel_to_snake(text: str) -> str:
+    """
+    Transforms camelCase to the snake_case.
+    """
+    return re.sub(r"(?<!^)(?=[A-Z])", "_", text).lower()
+
+
+def _code_example_exists(method_name: str, file_content: str):
+    return f"""test_{method_name}(""" in file_content
+
+
+def _create_hint(file_name: str, method_name: str) -> str:
+    """
+    Constructs a hint with code example.
+    """
+    return f"""
+        .. hint::
```

Comment:
Not sure about using hint for this. Couldn't it be a normal code block as a part of documentation? Or something else that is not a hint?

Reply:
It is a *hint* on how to use it though.
Plus hints have a nice, more contrasty color

Reply:
If you both think its okay, we can leave it as a hint. It just doesn't make 100% sense to me to have a hint that is just an example usage.

Reply:
I can rename it to `Example`

Reply:
![image](https://user-images.githubusercontent.com/61014013/214617446-2d838779-f713-4b8b-9c4a-6697ac4539c1.png)